pkg/puller/puller.go

```diff
@@ -354,76 +251,68 @@ func (p *Puller) histSyncWorker(ctx context.Context, peer swarm.Address, bin uin
-			}
-
-			// since we use bin context cancellation to cancel interval
-			// sync operations, the context can be expired here, causing us
-			// to try to send a message with an expired context, which is
-			// bound to fail.
-			ctxC, cancelC := context.WithTimeout(context.Background(), 10*time.Second)
-			defer cancelC()
-			if err := p.syncer.CancelRuid(ctxC, peer, ruid); err != nil {
-				loggerV2.Debug("histSyncWorker cancel ruid failed", "error", err)
-			}
-			return
+			p.metrics.HistWorkerErrCounter.Inc()
+			p.logger.Error(err, "histSyncWorker syncing interval failed", "peer_address", peer, "bin", bin, "cursor", cur, "start", s, "topmost", top)
+			sleep = true
 		}
+
 		err = p.addPeerInterval(peer, bin, s, top)
 		if err != nil {
 			p.metrics.HistWorkerErrCounter.Inc()
-			p.logger.Error(err, "could not persist interval for peer, quitting...", "peer_address", peer)
+			p.logger.Error(err, "histSyncWorker could not persist interval for peer, quitting...", "peer_address", peer)
 			return
 		}
+
 		loggerV2.Debug("histSyncWorker pulled", "bin", bin, "start", s, "topmost", top, "peer_address", peer)
 	}
 }
 
 func (p *Puller) liveSyncWorker(ctx context.Context, peer swarm.Address, bin uint8, cur uint64) {
-	loggerV2 := p.logger.V(2).Register()
+	loggerV2 := p.logger
 
 	defer p.wg.Done()
-	loggerV2.Debug("liveSyncWorker starting", "peer_address", peer, "bin", bin, "cursor", cur)
+	p.logger.Debug("liveSyncWorker starting", "peer_address", peer, "bin", bin, "cursor", cur)
 	from := cur + 1
+
+	sleep := false
+
 	for {
 		p.metrics.LiveWorkerIterCounter.Inc()
+
+		if sleep {
+			select {
+			case <-ctx.Done():
+				loggerV2.Debug("liveSyncWorker context cancelled", "peer_address", peer, "bin", bin, "cursor", cur)
+				return
+			case <-time.After(p.syncSleepDur):
+			}
+			sleep = false
+		}
+
 		select {
-		case <-p.quit:
-			loggerV2.Debug("liveSyncWorker quit on shutdown", "peer_address", peer, "bin", bin, "cursor", cur)
-			return
 		case <-ctx.Done():
 			loggerV2.Debug("liveSyncWorker context cancelled", "peer_address", peer, "bin", bin, "cursor", cur)
 			return
 		default:
 		}
-		top, ruid, err := p.syncer.SyncInterval(ctx, peer, bin, from, pullsync.MaxCursor)
-		if err != nil {
-			loggerV2.Debug("liveSyncWorker exit on sync error", "peer_address", peer, "bin", bin, "from", from, "error", err)
-			if ruid == 0 {
-				p.metrics.LiveWorkerErrCounter.Inc()
-			}
 
-			// since we use bin context cancellation to cancel interval
-			// sync operations, the context can be expired here, causing us
-			// to try to send a message with an expired context, which is
-			// bound to fail.
-			ctxC, cancelC := context.WithTimeout(context.Background(), 10*time.Second)
-			defer cancelC()
-			if err := p.syncer.CancelRuid(ctxC, peer, ruid); err != nil {
-				loggerV2.Debug("histSyncWorker cancel ruid failed", "error", err)
-			}
-			return
+		top, _, err := p.syncer.SyncInterval(ctx, peer, bin, from, pullsync.MaxCursor)
+		if err != nil {
```

Comment:
Same as the comment above. We should ideally look for particular errors to restart, if we get terminal errors like stream reset etc we should quit early.

Reply:
see comment above